rules.py

```diff
@@ -1010,6 +1010,52 @@ def check(self, element):
                             element.sourceline, element.tag, other_type_element.text))
 
 
+class GpUnitsCycleCheck(base.TreeRule):
+    """Check that GpUnits form a tree
+    Return error if cycle is found"""
+    nodes = {}
+    visited = {}
+    error_object = None
+    
+    def check(self):
+        for event, element in etree.iterwalk(self.election_tree):
+            tag = self.strip_schema_ns(element)
```

Comment:
This loop is building the dictionaries mapping nodes to their children and saying that we haven't visited any of these nodes, correct?